pkg/scheduler/framework/plugins/interpodaffinity/plugin.go

```diff
@@ -234,11 +258,36 @@ func (pl *InterPodAffinity) isSchedulableAfterNodeChange(logger klog.Logger, pod
 		return framework.Queue, err
 	}
 
+	// When queuing this Pod:
+	// - 1. A new node is added with the anti-affinity key, the pod may become schedulable.
+	// - 2. The original node does not have the anti-affinity key but the modified node does, the pod may become schedulable.
```

Comment:
Wait, I noticed the third scenario here..

```suggestion
	// - 2. The original node did not have the anti-affinity key but the modified node does, the pod may become schedulable.
	// - 3. The original node had the anti-affinity key but the modified node does not, the pod may become schedulable.
```

The third scenario example:

1. A pod has podantiaffinity for topologyKey: zone.
2. zone-a has three nodes and only node-1 has a Pod matching the pod's podantiaffinity. So, zone-a cannot get the Pod right now.
3. Then, node-1 lost zone-a label somehow; meaning node-1 doesn't belong to zone-a anymore and zone-a doesn't have any Pod matching the Pod's antiaffinity anymore. This label update to node-1 makes the pod schedulable to zone-a.

Reply:
This part seems to be more complicated than I thought. I don't know much about anitaffinity. I may need to spend some time on it.

Reply:
so we should  change to this right?
```go
                originalTopologyKey, originalHasKey := originalNode.Labels[term.TopologyKey]
		modifiedTopologyKey, modifiedHasKey := modifiedNode.Labels[term.TopologyKey]

		if !originalHasKey && modifiedHasKey {
			// Case 2: Original node does not have the key, but the modified node does.
			logger.V(5).Info("A node with a matching pod anti-affinity topologyKey was updated and it may make the pod schedulable",
				"pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
			return framework.Queue, nil
		}

		if originalHasKey && !modifiedHasKey {
			// Case 2: Original node has the key, but the modified node does not.
			logger.V(5).Info("A node with a matching pod anti-affinity topologyKey was updated and it may make the pod schedulable",
				"pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
			return framework.Queue, nil
		}
```

Reply:
If we need to consider topologykey disappearance in anti-affinity scenario, shouldn't we register for node framework.Delete events as well?

Also, am I correct that in affinity scenario we should requeue whenever a new node or new topologykey label appears, but in anit-affinity scenario requeue only when a topologykey label disappears or a node disappears?

In other words, IIUC in anti-affinity scenario, we don't need to requeue on new topologykey label appearance. If so, then the existing implementation is incorrect, as it should rather check the modifiedNode does not have the topologykey

Reply:
As far as I understand, if a pod has an anti-affinity label, when a node with this anti-affinity is deleted, the pod will most likely fail to be scheduled.

Reply:
This plugin is checking labels against other Pods labels affinity/anti-affinity, restricted the a given topology domain (specified by topologyKey key - like region, rack, zone or change and value - like us-east, us-west). Here we register for Node label changes only, which can in fact impact only the topology. A node can appear or disappear from a domain if a node or node label appears/disappears (originalHasKey != modifiedHasKey) or changes the domain if label value changes (us-east -> us-west when originalTopologyKey != modifiedTopologyKey).

So if we have an unschedulable Pod which cannot start in any region because in the only eligible ones there are Pods already running (triggering the anti-affinity), so the only case it can get schedule if the node with this anti-affinity Pod disappears. It can happen if the node changes the region (now we can schedule in us-east) or disappears from the domain (node is removed completely or the domain label (region) is removed)

Reply:
:astonished: To be honest, the scenarios of pod affinity antiaffinity and podtopologyspread are much more complicated than I thought....

Reply:
> This plugin is checking labels against other Pods labels affinity/anti-affinity, within a given topology domain (specified by the topologyKey key - like region, rack, zone or value - like us-east, us-west). Here we register for Node label changes only, which can in fact impact only the topology. A node can appear or disappear from a domain if a node or node label appears/disappears (originalHasKey != modifiedHasKey) or changes the domain if label value changes (us-east -> us-west when originalTopologyKey != modifiedTopologyKey).
> 
> So if we have an unschedulable Pod which cannot start in any region because in the only eligible ones there are Pods already running (triggering the anti-affinity), so the only case it can get schedule if the node with this anti-affinity Pod disappears. It can happen if the node changes the region (now we can schedule in us-east) or disappears from the domain (node is removed completely or the domain label (region) is removed)

According to the scenario you described, it is indeed necessary to add the Delete event

Reply:
> so the only case it can get schedule if the node with this anti-affinity Pod disappears

Is there any other case for Node/Delete except when it had a scheduled Pod with anti affinity? If not, won't the Pod/Delete event handle that case?

Reply:
> If we need to consider topologykey disappearance in anti-affinity scenario, shouldn't we register for node framework.Delete events as well?

Have the same impression as @macsko 's last comment. 
So, my description at the top of this thread about the third scenario is that node-1 has the Pod that is hated, but node-1 is moved to somewhere else with the hated Pod, which could make the Pod that is being scheduled schedulable. So, in this case, the hated Pod is disappeared from the zone without being deleted.

But, when it comes to Node/Delete event, before the Node is deleted, the hated Pod should be terminated and deleted, which triggers Pod/Delete QHint. 
So, I don't see the need to register Node/Delete event here.

Reply:
Ack, Pod/Delete should be seen indeed.

IIUC the case where adding node to the topology domain matters, is when this nodes opens a new region, where the anti-affinity pod could schedule.